waiter/src/waiter/util/http_utils.clj

```diff
@@ -76,9 +80,50 @@
         (.setUserAgentField client new-user-agent-field)))
     client))
 
-(defn determine-backend-protocol-version
+(defn- prepare-http2-transport
+  "Returns the HTTP/2 client transport."
+  [connection-timeout-ms]
+  (let [http2-client (HTTP2Client.)
+        http2-protocols (ArrayList. ["h2" "h2c"])]
+    (.setConnectTimeout http2-client connection-timeout-ms)
+    (.setProtocols http2-client http2-protocols)
+    (HttpClientTransportOverHTTP2. http2-client)))
+
+(defn prepare-http-clients
+  "Prepares and returns a map of HTTP clients for http/1 and http/2 requests."
+  [{:keys [conn-timeout user-agent] :as config}]
+  (let [http2-transport (prepare-http2-transport conn-timeout)]
+    {:http1-client (http-client-factory (cond-> config
+                                          user-agent (update :user-agent str ".http1")))
+     :http2-client (-> (cond-> config
+                         user-agent (update :user-agent str ".http2"))
+                       (assoc :transport http2-transport)
```

Comment:
ALPN works when the request can upgrade to HTTP/2 when we use this http/2 client.
```
(let [{:keys [http2-client]} (waiter.util.http-utils/prepare-http-clients {:conn-timeout 10000})]
  (-> (http/request http2-client {:url "https://www.google.com" :method :get :version "HTTP/1.1"})
      async/<!!
      clojure.pprint/pprint))
[C] ALPN protocols [h2, h2c] for 109539ef[SSLEngine[hostname=www.google.com port=443] SSL_NULL_WITH_NULL_NULL]
[C] ALPN protocol 'h2' selected by server for 109539ef[SSLEngine[hostname=www.google.com port=443] SSL_NULL_WITH_NULL_NULL]
... :request #object[org.eclipse.jetty.client.HttpRequest 0x61f16690 "HttpRequest[GET / HTTP/2.0]@61f16690"],
    :status 200}
```

When using a non-http/2 server, the connection hangs indefinitely:
```
(let [{:keys [http2-client]} (waiter.util.http-utils/prepare-http-clients {:conn-timeout 10000})]
  (-> (http/request http2-client {:url "https://www.dzone.com" :method :get :version "HTTP/1.1"})
      async/<!!
      clojure.pprint/pprint))
[C] ALPN protocols [h2, h2c] for 6a5ea8c1[SSLEngine[hostname=www.dzone.com port=443] SSL_NULL_WITH_NULL_NULL]
[C] ALPN not supported by server for 6a5ea8c1[SSLEngine[hostname=www.dzone.com port=443] SSL_NULL_WITH_NULL_NULL]
```

The [other example](https://www.programcreek.com/java-api-examples/?code=fstab/http2-examples/http2-examples-master/multiplexing-examples/jetty-client/src/main/java/de/consol/labs/h2c/examples/client/jetty/JettyClientExample.java) of [HTTP2Client  examples](https://www.programcreek.com/java-api-examples/?api=org.eclipse.jetty.http2.client.HTTP2Client) that I found required the `HttpVersion` to be set on the `MetaData.Request`.

Reply:
As per a [comment on Jetty GitHub on Feb 7, 2017](https://github.com/eclipse/jetty.project/issues/1308#issuecomment-277940891):

> HTTP2Client is a HTTP/2 only client.
> 
> HttpClient is the client that speak by default HTTP/1.1, although it can speak other protocols, see the documentation.
> 
> We have not implemented yet a mechanism, in the client, that can speak different protocols to different domains. We typically suggest to use different HttpClient instances configured each with a different transport, since most of the times the client knows what protocol a domain supports.